File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:

    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        seq = [] # not a str, because of multiple digit nums

        def DFS(root):
            nonlocal seq
            if not root: 
                seq.append("N")
                return

            seq.append(str(root.val))
            DFS(root.left)
            DFS(root.right)

        DFS(root)
        return ",".join(seq)
        
    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:       
        self.i = 0

        data = data.split(",")

        def DFS():
            if data[self.i] == "N":
                self.i += 1
                return None

            node = TreeNode(int(data[self.i]))

            self.i += 1

            node.left = DFS()
            node.right = DFS()

            return node

        return DFS()

File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) 
    {
        string seq;
        DFS_1(root, seq);
        return seq;
    }

    int DFS_1(TreeNode* root, string &seq)
    {
        if (!root)
        {
            seq.append("N,");
            return 0;
        }

        seq.append(to_string(root->val));
        seq.append(",");
        DFS_1(root->left, seq);
        DFS_1(root->right, seq);

        return 0;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) 
    {
        vector<string> seq = split(data, ',');
        int i = 0;
        return DFS_2(seq, i);
    }

    vector<string> split(const string &s, char delim) 
    {
        vector<string> elems;
        stringstream ss(s);
        string item;
        while (getline(ss, item, delim)) {
            elems.push_back(item);
        }
        return elems;
    }

    TreeNode* DFS_2(vector<string> seq, int &i)
    {
        if (seq[i] == "N")
        {
            i++;
            return NULL;
        }

        TreeNode* root = new TreeNode(stoi(seq[i]));
        i++;
        root->left = DFS_2(seq, i);
        root->right = DFS_2(seq, i);

        return root;
    }
};
